### PR TITLE
refactor(training): construct configuration before runtime initialization

### LINKED ISSUE #7324: Separate configuration preparation from runtime initialization
Parent initiative: #3554

## Goal

Make configuration preparation and runtime initialization explicit so the training entrypoints can construct and register `PretrainConfigContainer` before its downstream consumers run. Preserve current Megatron-LM behavior during this structural change.

## Context

`parse_and_validate_args()` currently applies checkpoint overrides and validation, then calls `set_global_variables(args)`. That registers args and constructs the tokenizer, microbatch calculator, loggers, timers, and other services. Entrypoints construct the model config and container afterward.

Tokenizer construction is a real prerequisite: it can derive `padded_vocab_size`, which GPT/hybrid model-config construction consumes. Moving all initialization after container construction would break that dependency.

## Scope

- Separate input parsing/checkpoint overrides/validation, tokenizer-dependent vocabulary resolution, container construction, and runtime-service initialization into explicit phases.
- Reuse `pretrain_cfg_container_from_args()` for container construction.
- Register the active container and expose `get_cfg()` with corresponding cleanup.
- Make tokenizer bootstrap explicit, preserve checkpoint-provided vocabulary precedence, and avoid constructing the tokenizer twice.
- Wire the in-scope training entrypoints through the explicit phases. Preserve the existing parse helper's behavior for callers outside the migration boundary through a compatibility wrapper where needed.
- Preserve required startup order, including early fault-tolerance monitoring, logging/timing, distributed setup, and restart behavior. Do not mechanically reorder services whose position affects behavior.
- Add focused startup tests and reusable baseline comparisons in the existing test structure.

This is the first initialization PR. The extracted helpers may still consume args; migrating their configuration ownership belongs to the follow-up subissue.

## Completion criteria

- Startup phases and their dependencies are explicit; the container is registered before migrated downstream consumers access it.
- Representative CLI configurations produce the same resolved model/config inputs and startup behavior as original MLM.
- Tests cover tokenizer-derived and explicitly/checkpoint-provided vocabulary sizes, single construction of runtime services, required initialization ordering, and cleanup/reinitialization.
- Existing training entrypoints and out-of-scope callers retain their supported behavior.

## Out of scope

- Making all initialization helpers or the full training loop config-authoritative.
- Porting all of `validate_args()` or migrating mutable progress to `TrainState` (#3555).
- New checkpoint metadata, callback APIs, or PEFT behavior.

Validation compares original MLM with refactored MLM.

### LINKED ISSUE #7325: Make training initialization and runtime services consume configs
Parent initiative: #3554

Depends on #7324 (explicit configuration preparation and initialization phases).

## Goal

Make the initialization path consume `PretrainConfigContainer` and its subconfigs so CLI-built and directly constructed configs initialize equivalent runtime services without relying on global args in the migrated helpers.

## Scope

- Migrate the initialization helpers extracted in #7324 and relevant `initialize.py`/`global_vars.py` consumers to configs or explicit config-derived inputs.
- Cover tokenizer construction and vocabulary resolution, distributed initialization, RNG, microbatch-calculator setup, logging/timers, profiling setup, rerun/straggler/fault-tolerance services, and associated async-checkpoint-worker setup.
- Complete the args-to-config mappings needed by these consumers. Resolve the fault-tolerance config schema against the parent initiative's mapping before implementing it; existing fault injection settings and FT/autoresume controls are not interchangeable.
- Move the validation and dependent-field derivation needed for initialization into appropriate config/shared helpers, with one implementation used by both input paths.
- Separate validation of requested launch topology before distributed setup from querying actual process-group sizes afterward. Replace runtime-derived args fields with helpers only where those helpers are valid at that lifecycle stage.
- Preserve tokenizer/vocabulary precedence, early monitoring, service ordering, restart, and cleanup behavior established by #7324.
- Keep compatibility args access localized to explicitly identified legacy entrypoint boundaries; migrated initialization helpers must not silently fall back to global args.

Model/DDP/optimizer construction and full-loop config migration follow in separate work. Shared initialization rules must remain consistent with those consumers during the transition.

## Completion criteria

- Initialization can be exercised from directly constructed configs with global `get_args()` unavailable in the migrated path.
- CLI and direct-config paths produce equivalent resolved tokenizer/vocabulary, runtime-service inputs, distributed topology, RNG initialization, and microbatch setup.
- Tests cover relevant translation rules, invalid combinations, pre/post-distributed derivation, startup ordering, and cleanup/reinitialization.
- Representative original MLM versus refactored MLM runs preserve initialization behavior and affected training behavior.
- Any remaining compatibility boundary is explicit and does not act as a second configuration authority.

## Out of scope

- Claiming the entire training loop is independent of args after this PR.
- Bulk migration of unrelated validation or model/optimizer/runtime consumers.
- `TrainState` ownership (#3555), new checkpoint metadata, callback APIs, or PEFT behavior.

### PR DESCRIPTION
**Pull request by cuichenx**
# What does this PR do?

Construct training configuration before initializing tokenizer, microbatch calculation, logging and other runtime services. This combines preparation and caller migration into one ordering refactor.

Fixes #7324
Fixes #7325

## Changes

- Make `parse_and_validate_args()` prepare/register normalized args without initializing runtime services. Keep checkpoint overrides and existing CLI validation in their original order; no old/new parser mode flag.
- Extract the existing service body into `initialize_runtime_services(args)`. Preserve the existing `set_global_variables()` contract for direct bootstrap callers.
- Update all 31 direct non-test parser callers: 10 training callers construct config before runtime setup; 21 args-only callers initialize services explicitly before distributed setup.
- Preserve known vocabulary, tokenizer-derived vocabulary and checkpoint override precedence. `resolve_tokenizer_vocab_size()` only binds the explicit resolved size; it does not initialize services or read/register globals.
- Add focused behavioral tests for parsing/runtime separation, service startup, vocabulary resolution and checkpoint precedence. Existing tests remain unchanged; source-layout guards and comparison tooling are kept outside this PR.

No `initialize_training_globals()` wrapper, new global full-config registry, FaultToleranceConfig, TrainState migration, checkpoint metadata or unrelated baseline fixes. Runtime services still consume normalized args in this step. Actual data sequence length remains distinct from model position capacity.

## Compatibility with #7344

This PR does not introduce or depend on its config registry. `training.py` and the training package exports remain unchanged. A separate composition check with #7344 passes its 12 tests plus two real-container/tokenizer/checkpoint round-trip tests.

## Validation

Latest rebase validation (2026-09-18): rebased onto frozen main `d564dd01de4549a4884be26b1d74e169798fd877`, including merged #7447. Current head: `f0148a3600818588406d8ca6437a8de1f8377f38`. `_set_train_state()` runs after telemetry inside `initialize_runtime_services()`, for both config-first and legacy args-only startup. The existing lifecycle test now checks both paths and verifies parsing does not create TrainState: **31 cases passed on one H100**. An AST comparison confirms the extracted runtime body exactly matches new main, including TrainState creation; its getter and cleanup are unchanged. Baseline/candidate full pre-commit reports the same 30 unrelated Black-modified files and identical Pylint diagnostics; isort passes and all 35 PR-file hashes remain unchanged. The 38-case equivalence matrix below has **not** been rerun on this new rebase.

Previous fixed-base validation (2026-09-17): baseline `cfa9e20d2658844dd84a07f2ccd34d7e62ab5552` versus rebased candidate `c92551aa6f63c64a5a495a94b784c19ad0966a8c`: **37 exact configuration/input-equivalence passes and one expected CLI rejection across all 38 cases**, including one-, two- and eight-GPU coverage. Zero differences or execution/capture errors. Same harness, fixtures and environment on both sides; no waived fields. [Detailed results and scope](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-seam-tests/tools/config_seam_tests/RESULTS.md).

The previously published head `f9add9679` added only a copyright comment to that tested candidate; both copyright checks passed. Full pre-commit on the pinned baseline and candidate reports identical unrelated Black/Pylint issues; all 35 PR-file hashes remain unchanged. These are short runtime/configuration checks, not a full convergence or performance gate.

Historical validation before the rebase, using baseline `54c62dfb6852e9218bf38590be9b90930bf7d589`:

- Focused original-MLM tests: 257 passed; refactored tests: 287 passed, including all 30 retained behavioral cases. The rerun excludes the removed source-layout/comparison tests.
- 8 H100 GPUs, TP2/PP2/DP2: original and refactored GPT match exactly for loss, learning rate, batch size and gradient norm across six fresh steps and four resumed steps, using `--use-checkpoint-args`.
- Independent complete-diff review: no actionable findings.
- The retained behavioral test passed the prior targeted Black/isort/Pylint/Ruff/mypy checks and is unchanged. The final all-files pre-commit rerun reports only preexisting failures in untouched files; no PR file was reformatted, and all 35 PR-file hashes match the checked source. Existing entrypoint import order and formatting are preserved.

Follow-up cleanup removes unrelated import sorting, call wrapping and whitespace. Imported bindings and non-import ASTs are unchanged across all 15 cleaned files; GPU training was not rerun for this cleanup. One small dynamic-inference import placement matches newer main to avoid a conflict; the three-way merge is clean.

The final test-scope cleanup removes two new test files and the comparison utility, preserving them privately for validation. Production code, existing tests, and the retained behavioral test are unchanged. The full PR was reviewed again with no new actionable correctness finding.

Remaining scope: broader artifact-dependent/model-family coverage and full convergence/performance validation are not established by these focused tests. No fully green suite or general args-free/direct-config training is claimed.

## Pre-checks

- [x] Relevant unit tests added.
- [x] Fresh/resume behavior compared against original MLM.
- [x] Type hints and API docstrings added for new helpers.
- [x] Repository autoformat check run.
- [ ] Complete broader validation before marking ready for review.
- [ ] PR author review/sign-off on the final scope and remaining coverage.

No GPU kernels, dependencies or CI workflows changed.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -107,9 +113,11 @@ def parse_and_validate_args(extra_args_provider=None, ignore_unknown_args=False,
     else:
         validate_args(args, args_defaults)
 
-    # set global args, build tokenizer, and set adlr-autoresume,
-    # tensorboard-writer, and timers.
-    set_global_variables(args)
+    set_args(args)
```
Comment (cuichenx) [inline megatron/training/arguments.py]: do we not need _ensure_var_is_not_initialized before set_global_variables?

File: tests/unit_tests/training/test_training_entrypoint_ordering.py
Comment (cuichenx) [inline tests/unit_tests/training/test_training_entrypoint_ordering.py]: we can remove this file, this test seems forced

File: pretrain_gpt.py [hunk 2]
```diff
@@ -570,10 +572,14 @@ def get_embedding_ranks(pp_ranks: List[int]):
...
-        model_cfg = gpt_config_from_args(args, model_config_cls=ModelOptModelConfig)
+        model_cfg = gpt_config_from_args(
+            args, model_config_cls=ModelOptModelConfig, vocab_size_from_tokenizer=True
+        )
     else:
-        model_cfg = gpt_config_from_args(args)
+        model_cfg = gpt_config_from_args(args, vocab_size_from_tokenizer=True)
     full_config = pretrain_cfg_container_from_args(args, model_cfg)
+    initialize_runtime_services(args)
+    resolve_tokenizer_vocab_size(full_config, args.padded_vocab_size)
```
Comment (maanug-nv) [inline pretrain_gpt.py:582]: i think `resolve_tokenizer_vocab_size(full_config, tokenizer.vocab_size)` will be more reliable in cases where '--padded-vocab-size' was not specified.

File: megatron/training/argument_utils.py [hunk 3]
```diff
@@ -447,14 +433,43 @@ def _default_config_from_args(cls: type, args: Namespace, return_instance: bool
         return kwargs
 
 
+def resolve_tokenizer_vocab_size(cfg: PretrainConfigContainer, padded_vocab_size: int | None) -> None:
```
Comment (cuichenx) [inline megatron/training/argument_utils.py:450]: this will be moved to the config container's `finalize()` function afterwards

File: pretrain_gpt.py (hunk 2 above)
Comment (cuichenx) [reply pretrain_gpt.py:582]: discussed offline, this line should take padded vocab size which is always populated in build_tokenizer() even if user does not specify this flag

**Comment by cuichenx** [comment]
Validation rerun completed for **`f0148a3600818588406d8ca6437a8de1f8377f38`**, compared against frozen main `d564dd01de4549a4884be26b1d74e169798fd877`.

- Suite: [fork-only configuration-seam tests](https://github.com/cuichenx/Megatron-LM/tree/cf8495a52529164ead3f41a385061e1743ead86e/tools/config_seam_tests), unchanged capture code and 38-case manifest.
- **38/38 expected outcomes:** 37 exact configuration/consumer-input matches plus the expected `invalid_batch_conflict` CLI rejection. **Zero differences or execution/capture errors.**
- Coverage: 27 builder cases and 10 short runtime cases on 1/2/8 H100s, including checkpoint resume, recompute, fine-tuning, frozen-vision VLM, and TP/PP/DP/CP/EP combinations.
- All 13 harness self-tests passed in each allocation.

[Detailed results](https://github.com/cuichenx/Megatron-LM/blob/cf8495a52529164ead3f41a385061e1743ead86e/tools/config_seam_tests/RESULTS.md). This validates captured configuration/input equivalence and short-run completion, not full convergence or performance equivalence.

**Review by maanug-nv — APPROVED** [review]

File: examples/post_training/modelopt/offline_feature_extract.py [hunk 4]
```diff
@@ -43,11 +44,12 @@ def extract_feature(dataset, model, output_dir, idx_start, idx_end):
             torch.distributed.barrier()
 
 if __name__ == "__main__":
-    parse_and_validate_args(extra_args_provider=add_extract_args, args_defaults={
+    args = parse_and_validate_args(extra_args_provider=add_extract_args, args_defaults={
```
Comment (tdene) [inline examples/post_training/modelopt/offline_feature_extract.py:47]: It looks like this script never imports `parse_and_validate_args`.

File: examples/multimodal/train.py [hunk 5]
```diff
@@ -389,6 +391,8 @@ def write_online_eval_to_tensorboard(data, iteration, writer, walltime=None):
         args_defaults={'tokenizer_type': 'GPT2BPETokenizer'},
     )
     full_config = pretrain_cfg_container_from_args(args)
+    initialize_runtime_services(args)
+    resolve_tokenizer_vocab_size(full_config, args.padded_vocab_size)
     pretrain(
```
Comment (tdene) [inline examples/multimodal/train.py:396]: `pretrain` is supposed to take `cfg_container` as its first argument. Since these are all positional arguments, this script is broken, with all arguments offset by one.

File: megatron/training/argument_utils.py [hunk 6]
```diff
@@ -493,15 +522,22 @@ def gpt_config_from_args(
         kwargs["vocab_size"] = args.padded_vocab_size
```
Comment (tdene) [inline megatron/training/argument_utils.py]: With `--yaml-cfg`, args are read directly from the yaml. Without `--yaml-cfg`, the tokenizer is built inside `parse_and_validate_args` and `padded_vocab_size` is built there. But since that doesn't run with `--yaml-cfg`, we can end up with an error when the config holds `pad_vocab_size: true` but does not define `padded_vocab_size`. That's a regression from the legacy non-yaml path.

File: megatron/training/argument_utils.py [hunk 7]
```diff
@@ -447,14 +447,43 @@ def _default_config_from_args(cls: type, args: Namespace, return_instance: bool
...
+        cfg.model.should_pad_vocab = False
+
+
 def gpt_config_from_args(
-    args: Namespace, config: TransformerConfig | None = None, model_config_cls: type = GPTModelConfig
+    args: Namespace,
+    config: TransformerConfig | None = None,
+    model_config_cls: type = GPTModelConfig,
+    *,
+    vocab_size_from_tokenizer: bool = False,
```
Comment (tdene) [inline megatron/training/argument_utils.py]: Another one related to the fact that config is initialized before the tokenizer. This one is if the `--yaml-cfg` settings is not used.

A `False` default for `vocab_size_from_tokenizer` tries to replicate the pre-existing logic, but it's dangerous now that the tokenizer is not built until after the config is completed.

If the caller does not set `padded_vocab_size` in the args, and also fails to set `vocab_size_from_tokenizer=True` to change this default setting, the code tries to default to `vocab_size` instead of `padded_vocab_size` causing a mismatch.

**Review by tdene — APPROVED** [review]

File: examples/multimodal/train.py (hunk 5 above)
Comment (cuichenx) [reply examples/multimodal/train.py:396]: thanks for noticing this

File: megatron/training/argument_utils.py (hunk 6 above)
Comment (maanug-nv) [reply megatron/training/argument_utils.py]: tokenizer is not built in `parse_and_validate_args` ([code](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-first-initialization/megatron/training/arguments.py#L89-L123)) anymore for either path, right?

In both yaml and non-yaml cases, `padded_vocab_size` will be set [here](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-first-initialization/megatron/core/tokenizers/utils/build_tokenizer.py#L115-L116) if user set `pad_vocab_size`. Then this [resolve](https://github.com/cuichenx/Megatron-LM/blob/chcui/omar/config-first-initialization/pretrain_gpt.py#L582C1-L583C1) call will address updating vocab size settings in the config, after tokenizer was built.

What am I missing?

File: megatron/training/argument_utils.py (hunk 7 above)
Comment (cuichenx) [reply megatron/training/argument_utils.py]: As long as user sets `pad_vocab_size` in the yaml, `padded_vocab_size` will still be set in `build_tokenizer()`.
Also i'm setting vocab_size_from_tokenizer=True as the default now the equivalent before this refactor was so use vocab size from tokenizer.

**Review by Phlip79 — APPROVED** [review]

File: megatron/training/argument_utils.py (hunk 6 above)
Comment (tdene) [reply megatron/training/argument_utils.py]: `1e2ddf8` fixed this already, but

You're right, I was in a rush to leave this review before my vacation and miswrote pretty badly. The tokenizer is not built inside `parse_and_validate_args`, but inside `initialize_runtime_services`.

My problem was that if the user does not set `pad_vocab_size`, `gpt_config_from_args` was calling `args.padded_vocab_size` directly, and it didn't have a default on the YAML path. I could have kept it simpler; it got fixed with just a `padded_vocab_size = getattr(args, "padded_vocab_size", None)`.
